src/satosa/micro_services/idp_metadata_attribute_store.py

```diff
@@ -0,0 +1,229 @@
+            idp_config.update(config[idp])
+
+            self.config[idp] = idp_config
+
+        satosa_logging(logger, logging.INFO, "IdP Metadata Attribute Store microservice initialized", None)
+
+    def _first_lang_element_text(self, elements, lang='en'):
+        """
+        Loop over the list representing XML elements that contain text and find
+        the first text value for the input lang where 'en' or English is the
+        default lang.
+
+        Each item in the list is a dictionary with keys
+
+            __class__
+            lang
+            text
+
+        as expected from the metadata returned for an entity by the MetadataStore
+        class from pysaml2.
+
+        If no element has the input lang then return the text from the first 
+        element.
+
+        If no element has text then return an empty string.
+        """
+        for e in elements:
+            if lang in e:
+                if 'text' in e:
+                    return e['text']
+
+        for e in elements:
+            if 'text' in e:
+                return e['text']
+
+        return ''
+
+    def process(self, context, data):
+        """
+        Default interface for microservices. Process the input data for
+        the input context.
+        """
+        self.context = context
+
+        # Find the entityID for the IdP that issued the assertion.
+        try:
+            idp_entity_id = data.to_dict()['auth_info']['issuer']
+        except KeyError as err:
+            satosa_logging(logger, logging.ERROR, "Unable to determine the entityID for the IdP issuer", context.state)
+            return super().process(context, data)
+
+        # Get the configuration for the IdP.
+        if idp_entity_id in self.config.keys():
+            config = self.config[idp_entity_id]
+        else:
+            config = self.config['default']
+
+        satosa_logging(logger, logging.DEBUG, "Using config {}".format(config), context.state)
+
+        # Ignore this IdP if so configured.
+        if config['ignore']:
+            satosa_logging(logger, logging.INFO, "Ignoring IdP {}".format(idp_entity_id), context.state)
+            return super().process(context, data)
+
+        # Get the metadata store the SP for the proxy is using. This
+        # will be an instance of the class MetadataStore from mdstore.py
+        # in pysaml2.
+        metadata_store = context.internal_data['metadata_store']
+
+        # Get the metadata for the IdP.
+        try:
+            metadata = metadata_store[idp_entity_id]
+        except Exception as err:
+            satosa_logging(logger, logging.ERROR, "Unable to retrieve metadata for IdP {}".format(idp_entity_id), context.state)
+            return super().process(context, data)
+
+        satosa_logging(logger, logging.DEBUG, "Metadata for IdP {} is {}".format(idp_entity_id, metadata), context.state)
+
+        # Find the mdui:DisplayName for the IdP if so configured.
```

Comment:
I would split the processing of the attributes into different classes. This may be a bit of over-engineering but it allows to extend the mechanism by only creating a new class with just the appropriate _find_ handler and mapping it to the factory. It also makes code reusable. 

```py
from abc import ABC

# abstract class acts like an interface
class GenericAttribute(ABC): # not sure about the name
  @abstractmethod
  def find(self, metadata, lang, *argv):
    raise NotImplementedError

# organization_display_name and organization_name use the same mechanism for find()
class GenericName(GenericAttribute, ABC):
  def find(self, metadata, lang, *argv):
    org_display_name_elements = metadata['organization'][self.attr_name]
    value = self._first_lang_element_text(org_display_name_elements, lang)
    return value

class OrganizationDisplayName(GenericName):
  attr_name = 'organization_display_name'

class OrganizationName(GenericName):
  attr_name = 'organization_name'

class DisplayName(GenericAttribute):
  attr_name = 'display_name'

  def find(self, metadata, lang, *argv):
    extensions = metadata['idpsso_descriptor'][0]['extensions']['extension_elements']
    for e in extensions:
      if e['__class__'] == 'urn:oasis:names:tc:SAML:metadata:ui&UIInfo':
        display_name_elements = e[self.attr_name]
        value = self._first_lang_element_text(display_name_elements, lang)
        break
    return value


class AttributeFactory:
  factory = {
    'organization_display_name': OrganizationDisplayName,
    'organization_name': OrganizationName,
    'display_name': DisplayName,
  }

  def create(attr_type):
    try:
      attr_class = factory[attr_type]
    except KeyError as err:
      msg_err = "No handler for attribute: {}".format(attr_type)
      raise IdpMetadataAttributeStoreError(msg_err) from err
    else:
      return attr_class()

def process(..):
  ...
  for attr in config.keys():
    attribute_handler = AttributeFactory.create(attr)
    value = attribute_handler.find(metadata, lang)
    if value is not None:
      data.attributes[config[attr]['internal_attribute_name']] = value
      msg_dbg = "attribute {} is {}".format(attr, value)
      satosa_logging(logger, logging.DEBUG, msg_dbg, context.state)
```